### DISCUSSION TITLE
[@types/jest] spyOn doesn't work well with TypeScript conditional types

### DISCUSSION THREAD

**Discussion by ghost** [DefinitelyTyped/DefinitelyTyped#55380] [Issues with a @types package] [answered]
Hello,

Can't find any information on this so perhaps I'm doing something wrong?

We have a function called `useFeatures`, which has conditional types, as per the documentation: https://www.typescriptlang.org/docs/handbook/2/conditional-types.html

```
function useFeatures(feature: Feature): boolean;
function useFeatures(feature: Feature[]): boolean[];
function useFeatures(feature: Feature | Feature[]): boolean | boolean[] {
  const { features } = useContext(FeaturesContext);

  if (Array.isArray(feature)) {
    return feature.map((feat) => isFeatureOn(features, feat)) as boolean[];
  }

  return isFeatureOn(features, feature) as boolean;
}
```

and when trying to mock this, we do the following:

```
    import * as features from './features';

    jest.spyOn(features, 'useFeatures').mockImplementation(() => false);
```

we can probably mock this in a better way, but the TypeScript error we're seeing in the above is the following:

> Type 'boolean' is not assignable to type 'boolean[]'.ts(2322)

Is this an issue with the library types, or do we need to pass a generic?

[a comment by an automated account is not shown]

**Comment by domdomegg** [comment, marked as answer]
Hi @Ed-Ed, this is a limitation of the ReturnType helper in TypeScript - see https://github.com/Microsoft/TypeScript/issues/24275

You may find this StackOverflow answer helpful for your case (add a final overload as the more generic function type):
https://stackoverflow.com/a/52760599/5352227

**Comment by aayankhanali258090-creator** [comment]
Free fire diamond UID 7772004079
